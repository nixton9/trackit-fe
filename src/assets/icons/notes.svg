<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><defs><style>.cls-1{fill:#fff;}</style></defs><title>Notes</title><g id="notes"><path class="svg-fill" d="M352,80V64a16,16,0,0,0-32,0V80H272V64a16,16,0,0,0-32,0V80H192V64a16,16,0,0,0-32,0V80a80.09,80.09,0,0,0-80,80V384a80.09,80.09,0,0,0,80,80H352a80.09,80.09,0,0,0,80-80V160A80.09,80.09,0,0,0,352,80Zm48,304a48.05,48.05,0,0,1-48,48H160a48.05,48.05,0,0,1-48-48V160a48.05,48.05,0,0,1,48-48v16a16,16,0,0,0,32,0V112h48v16a16,16,0,0,0,32,0V112h48v16a16,16,0,0,0,32,0V112a48.05,48.05,0,0,1,48,48Z"/><path class="svg-fill" d="M336,208H176a16,16,0,0,0,0,32H336a16,16,0,0,0,0-32Z"/><path class="svg-fill" d="M336,267H176a16,16,0,0,0,0,32H336a16,16,0,0,0,0-32Z"/><path class="svg-fill" d="M336,326H176a16,16,0,0,0,0,32H336a16,16,0,0,0,0-32Z"/></g></svg>