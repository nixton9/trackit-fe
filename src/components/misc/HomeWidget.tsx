import React, { ReactElement } from 'react'
import { Styled } from '../../styles/HomeWidget'
import { Link } from 'react-router-dom'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'

export enum HomeWidgetTypes {
  Notes = 'Notes',
  Tasks = 'Tasks',
  Habits = 'Habits',
  Expenses = 'Expenses'
}

interface HomeWidgetProps {
  type: HomeWidgetTypes
  value: string
  label: string
  url: string
  icon: ReactElement
}

const HomeWidget: React.FC<HomeWidgetProps> = ({
  type,
  value,
  label,
  url,
  icon
}) => {
  return (
    <Link to={`/${type.toLowerCase()}`}>
      <Styled.Widget type={type}>
        <Styled.WidgetHeader>
          <Styled.WidgetHeader__Title>{type}</Styled.WidgetHeader__Title>
          <Styled.WidgetHeader__Icon>{icon}</Styled.WidgetHeader__Icon>
        </Styled.WidgetHeader>

        <Styled.WidgetInfo>
          <Styled.WidgetInfo_Value
            length={type === HomeWidgetTypes.Expenses ? value.length : 0}
          >
            {value}
            {type === HomeWidgetTypes.Expenses && '$'}
          </Styled.WidgetInfo_Value>
          <Styled.WidgetInfo_Label>{label}</Styled.WidgetInfo_Label>
        </Styled.WidgetInfo>

        <Styled.WidgetPlus>
          <PlusIcon />
        </Styled.WidgetPlus>
      </Styled.Widget>
    </Link>
  )
}

export default HomeWidget
