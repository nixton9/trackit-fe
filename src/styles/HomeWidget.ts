import styled from 'styled-components'
import { HomeWidgetTypes } from '../components/misc/HomeWidget'
import wave1 from '../assets/wave1.svg'
import wave2 from '../assets/wave2.svg'
import wave3 from '../assets/wave3.svg'
import wave4 from '../assets/wave4.svg'

type WidgetProps = {
  type: HomeWidgetTypes
}

type WidgetInfo__ValueProps = {
  length: number
}

const Widget = styled.div<WidgetProps>`
  position: relative;
  color: ${props => props.theme.white};
  border-radius: ${props => props.theme.bigBorderRadius};
  width: 12rem;
  height: 12.5rem;
  padding: ${props => props.theme.spacingXXS};
  background: ${props =>
    props.type === HomeWidgetTypes.Notes
      ? `url(${wave1}), ${props.theme.blueGradient}`
      : props.type === HomeWidgetTypes.Tasks
      ? `url(${wave2}), ${props.theme.purpleGradient}`
      : props.type === HomeWidgetTypes.Habits
      ? `url(${wave3}), ${props.theme.pinkGradient}`
      : `url(${wave4}), ${props.theme.greenGradient}`};
  background-size: cover;

  .svg-fill {
    fill: ${props => props.theme.white};
  }
  .svg-stroke {
    stroke: ${props => props.theme.white};
  }

  &:before {
    content: '';
    width: 82%;
    height: 93%;
    position: absolute;
    top: 12%;
    left: 0;
    right: 0;
    margin: 0 auto;
    filter: blur(31px);
    opacity: 70%;
    border-radius: ${props => props.theme.mainBorderRadius};
    background: ${props =>
      props.type === HomeWidgetTypes.Notes
        ? props.theme.blueGradient
        : props.type === HomeWidgetTypes.Tasks
        ? props.theme.purpleGradient
        : props.type === HomeWidgetTypes.Habits
        ? props.theme.pinkGradient
        : props.theme.greenGradient};
    z-index: -1;
  }
`

const WidgetHeader = styled.div`
  display: flex;
  justify-content: space-between;
`

const WidgetHeader__Title = styled.h3`
  font-weight: ${props => props.theme.fontBold};
  font-size: 1.25rem;
  line-height: 1.4rem;
`

const WidgetHeader__Icon = styled.span`
  svg {
    width: 2.2rem;
  }
`

const WidgetInfo = styled.div`
  margin-top: ${props => props.theme.spacingXXS};
  padding-left: 0.5rem;
  overflow: hidden;
`

const WidgetInfo_Value = styled.h2<WidgetInfo__ValueProps>`
  font-weight: ${props => props.theme.fontExtraLight};
  font-size: ${props =>
    props.length < 3 ? '4.8rem' : props.length < 4 ? '3.6rem' : '3rem'};
  line-height: 4.5rem;
`

const WidgetInfo_Label = styled.p`
  font-weight: ${props => props.theme.fontRegular};
  font-size: 0.75rem;
  padding-left: 0.2rem;
`

const WidgetPlus = styled.span`
  position: absolute;
  right: 0.7rem;
  bottom: 0.5rem;

  svg {
    width: 1.5rem;
  }
`

export const Styled = {
  Widget,
  WidgetHeader,
  WidgetHeader__Title,
  WidgetHeader__Icon,
  WidgetInfo,
  WidgetInfo_Value,
  WidgetInfo_Label,
  WidgetPlus
}
