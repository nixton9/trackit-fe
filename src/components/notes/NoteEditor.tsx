import React, { Dispatch, SetStateAction } from 'react'
import ReactQuill from 'react-quill'
import styled from 'styled-components'
import { device } from '../../styles/theme'
import 'react-quill/dist/quill.snow.css'

type EditorContainerProps = {
  readMode?: boolean
  showEditor?: boolean
}

const EditorContainer = styled.div<EditorContainerProps>`
  margin-top: ${props => props.theme.spacingS};

  .ql-toolbar {
    border: 2px solid ${props => props.theme.white};
    border-radius: ${props => props.theme.smallBorderRadius};
    display: ${props => (props.showEditor ? 'block' : 'none')};

    @media ${device.mobile} {
      .ql-formats {
        margin-right: 0.5rem;
      }
    }
  }

  .ql-container {
    min-height: ${props => (props.readMode ? 'unset' : '40vh')};
    max-height: ${props => (props.readMode ? 'unset' : '55vh')};
    overflow-y: auto;
    border: none;
    margin: ${props => props.theme.spacingXXS} 0;
    font-family: ${props => props.theme.fontFamily};
  }

  .ql-editor {
    padding: ${props => props.theme.spacingS};
  }

  .ql-snow .ql-stroke {
    stroke: ${props => props.theme.white};
  }

  .ql-snow .ql-picker-label::before {
    color: ${props => props.theme.white};
  }

  .ql-editor p {
    color: ${props => props.theme.white};
    font-size: 1.6rem;
    line-height: 3rem;
    font-weight: ${props =>
      props.readMode ? props.theme.fontRegular : props.theme.fontLight};
  }

  .ql-editor.ql-blank::before {
    left: ${props => props.theme.spacingS};
    color: ${props => props.theme.white};
    font-size: 1.7rem;
    font-weight: ${props => props.theme.fontLight};
    font-style: normal;
    opacity: 0.8;
  }
`

type NoteEditorProps = {
  value: string
  placeholder?: string
  readMode?: boolean
  showEditor?: boolean
  setValue?: Dispatch<SetStateAction<string>>
}

export const NoteEditor: React.FC<NoteEditorProps> = ({
  value,
  placeholder,
  readMode,
  showEditor = true,
  setValue
}) => {
  return (
    <EditorContainer readMode={readMode} showEditor={showEditor}>
      <ReactQuill value={value} onChange={setValue} placeholder={placeholder} />
    </EditorContainer>
  )
}
