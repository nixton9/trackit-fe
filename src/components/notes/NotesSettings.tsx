import React, { useState, useEffect, Dispatch, SetStateAction } from 'react'
import Drawer from '../misc/Drawer'
import Tag from './Tag'
import { TagEditor } from './TagEditor'
import { SelectMenu } from '../misc/SelectMenu'
import { NotificationTypes, notificationState } from '../misc/Notification'
import { alertState } from '../misc/Alert'
import { Styled } from '../../styles/Settings.styles'
import { TAGS, NOTES } from '../../utils/queries'
import { useToggleElement } from '../../utils/useToggleElement'
import { ReactComponent as SettingsIcon } from '../../assets/icons/settings.svg'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'
import { NoteTag } from '../../utils/ModuleTypes'
import { SortBySettings } from '../../utils/SettingsTypes'
import { useSetRecoilState } from 'recoil'
import { gql, useMutation, useQuery } from '@apollo/client'
import { useHistory } from 'react-router-dom'

const DELETE_TAG = gql`
  mutation DeleteTag($id: ID!) {
    deleteTag(id: $id) {
      id_tag
    }
  }
`

const sortByOptions = [
  { val: SortBySettings.DATE, label: 'Date' },
  { val: SortBySettings.ALPHABETICAL, label: 'Alphabetical Order' }
]

type NotesSettingsProps = {
  tags: NoteTag[]
  sortBy: string
  setSortBy: Dispatch<SetStateAction<SortBySettings>>
}

const NotesSettings: React.FC<NotesSettingsProps> = ({
  tags,
  sortBy,
  setSortBy
}) => {
  const [showTagEditor, setShowTagEditor] = useState(false)
  const [activeTag, setActiveTag] = useState<NoteTag | null>(null)

  const setNotification = useSetRecoilState(notificationState)
  const setAlert = useSetRecoilState(alertState)

  const history = useHistory()

  const [open, setOpen, overlayEl] = useToggleElement(() =>
    setShowTagEditor(false)
  )

  const { refetch: refetchTags } = useQuery(TAGS, {
    fetchPolicy: 'network-only'
  })
  const { refetch: refetchNotes } = useQuery(NOTES)

  const [deleteTag] = useMutation(DELETE_TAG)

  const handleSortByChange = (e: React.ChangeEvent<HTMLSelectElement>) =>
    setSortBy(e.target.value as SortBySettings)

  const handleTagClick = (tag: NoteTag) => {
    setActiveTag(tag)
    setShowTagEditor(true)
  }

  const handlePlusClick = () => {
    setActiveTag(null)
    setShowTagEditor(true)
  }

  const handleDeleteTag = (tagId: string | number) => {
    deleteTag({
      variables: {
        id: tagId
      }
    })
      .then(res => {
        setNotification({
          text: 'Tag was deleted',
          type: NotificationTypes.Success
        })
        refetchTags()
        refetchNotes()
      })
      .catch(err =>
        setNotification({
          text: 'There was a problem, please try again',
          type: NotificationTypes.Error
        })
      )
  }

  const handleDeleteTagConfirm = (tagId: string | number) => {
    setShowTagEditor(false)
    setAlert({
      text: 'This tag will be removed.',
      onConfirm: () => handleDeleteTag(tagId)
    })
  }

  useEffect(() => {
    history.push(`#`)

    window.onpopstate = (e: any) => {
      setOpen(false)
    }
  }, [history, setOpen])

  return (
    <>
      <SettingsIcon
        className="settings-icon"
        onClick={() => setOpen(true)}
        data-test-id="notes-settings-icon"
      />

      <Drawer title="Settings" open={open} overlayRef={overlayEl}>
        <>
          <Styled.SettingsBlock>
            <Styled.SettingsBlock__Label htmlFor="sort-by">
              Sort by
            </Styled.SettingsBlock__Label>
            <Styled.SettingsBlock__Input>
              <SelectMenu
                id="sort-by"
                value={sortBy}
                onChange={handleSortByChange}
                options={sortByOptions}
              />
            </Styled.SettingsBlock__Input>
          </Styled.SettingsBlock>

          <Styled.SettingsBlock>
            <Styled.SettingsBlock__Label>Tags</Styled.SettingsBlock__Label>
            <Styled.SettingsBlock__Tags>
              {tags &&
                tags.map(tag => (
                  <Tag
                    key={tag.id}
                    id={tag.id}
                    name={tag.name}
                    color={tag.color}
                    onClick={() => handleTagClick(tag)}
                    onDelete={() => handleDeleteTagConfirm(tag.id)}
                  />
                ))}
              <Styled.SettingsBlock__Icon
                onClick={handlePlusClick}
                className="mbl-click"
                data-test-id="tags-add-icon"
              >
                <PlusIcon />
              </Styled.SettingsBlock__Icon>
            </Styled.SettingsBlock__Tags>
          </Styled.SettingsBlock>

          {showTagEditor && (
            <TagEditor
              tag={activeTag}
              closeEditor={() => setShowTagEditor(false)}
            />
          )}
        </>
      </Drawer>
    </>
  )
}

export default NotesSettings
