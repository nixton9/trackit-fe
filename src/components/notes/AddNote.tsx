import React, { useState } from 'react'
import { NoteEditor } from './NoteEditor'
import { AddSubmitButton } from '../misc/Add'
import { LoadingSpinner } from '../misc/LoadingSpinner'
import { NotificationTypes, notificationState } from '../misc/Notification'
import { Styled } from '../../styles/Add.styles'
import { TagsInput, Tag } from './TagsInput'
import { NOTES } from '../../utils/queries'
import theme from '../../styles/theme'
import { DrawerAddModuleProps } from '../misc/Add'
import { ReactComponent as NotesIcon } from '../../assets/icons/notes.svg'
import { gql, useMutation, useQuery } from '@apollo/client'
import { useSetRecoilState } from 'recoil'
import 'react-quill/dist/quill.snow.css'

const CREATE_NOTE = gql`
  mutation CreateNote($title: String!, $content: String!) {
    createNote(title: $title, content: $content) {
      id_note
      title_note
      content_note
    }
  }
`

const ADD_TAG_TO_NOTE = gql`
  mutation AddTagToNote($note: ID!, $tag: ID!) {
    addTagToNote(note: $note, tag: $tag) {
      note_id
    }
  }
`

const CREATE_TAG = gql`
  mutation CreateTag($name: String!, $color: String!) {
    createTag(name: $name, color: $color) {
      id_tag
    }
  }
`

type CreateNoteData = {
  createNote: {
    id_note: string
    title_note: string
    content_note: string
  }
}

type CreateTagData = {
  createTag: {
    id_tag: string
  }
}

type AddTagToNotData = {
  addTagToNote: {
    note_id: string
  }
}

const AddNote: React.FC<DrawerAddModuleProps> = ({ closeModal }) => {
  const [title, setTitle] = useState('')
  const [content, setContent] = useState('')
  const [tags, setTags] = useState<Tag[] | []>([])

  const setNotification = useSetRecoilState(notificationState)

  const { refetch: refetchNotes } = useQuery(NOTES)
  const [addTagToNote] = useMutation<AddTagToNotData>(ADD_TAG_TO_NOTE)
  const [createTag] = useMutation<CreateTagData>(CREATE_TAG)
  const [createNote, { loading }] = useMutation<CreateNoteData>(CREATE_NOTE, {
    variables: { title: title, content: content }
  })

  // If there are tags, we'll add them to the note. If a tag is new we need to create it first
  // and only then assign it to the note
  const handleSubmit = (e: React.FormEvent<EventTarget>) => {
    e.preventDefault()
    if (title) {
      createNote()
        .then(results => {
          if (results && results.data && results.data.createNote) {
            const noteId = results.data.createNote.id_note
            tags.forEach(tag => {
              if (isNaN(Number(tag.id))) {
                createTag({
                  variables: { name: tag.text, color: pickRandomColor() }
                })
                  .then(res => {
                    res.data &&
                      attachTagToNote(noteId, res.data.createTag.id_tag)
                  })
                  .catch(err => console.log(err))
              } else {
                attachTagToNote(noteId, tag.id)
              }
            })
            setNotification({
              text: `New note created '${title}'`,
              type: NotificationTypes.Success
            })
            refetchNotes()
            closeModal()
          }
        })
        .catch(err =>
          setNotification({
            text: 'There was a problem, please try again',
            type: NotificationTypes.Error
          })
        )
    } else {
      setNotification({
        text: `You need to insert a title for your note`,
        type: NotificationTypes.Error
      })
    }
  }

  const attachTagToNote = (note: string, tag: string) => {
    addTagToNote({
      variables: { note: note, tag: tag },
      refetchQueries: () => [
        {
          query: NOTES
        }
      ]
    }).catch(err => console.log(err.message))
  }

  const pickRandomColor = () => {
    const { categories } = theme
    const randomColor = Object.keys(categories)[
      Math.floor(Math.random() * Object.keys(categories).length)
    ]
    return (categories as any)[randomColor]
  }

  const focusTagInput = () => {
    const input = document.querySelector('.ReactTags__tagInputField') as any
    input && input.focus()
  }

  return loading ? (
    <Styled.AddLoading>
      <LoadingSpinner />
    </Styled.AddLoading>
  ) : (
    <>
      <form onSubmit={handleSubmit}>
        <Styled.AddInputWrapper>
          <Styled.AddInput__Label>Title</Styled.AddInput__Label>
          <Styled.AddInput
            type="text"
            placeholder="Ex: Groceries list"
            value={title}
            onChange={e => setTitle(e.target.value)}
          />
        </Styled.AddInputWrapper>

        <Styled.AddEditor>
          <NoteEditor
            value={content}
            setValue={setContent}
            placeholder="Start writing your note here"
          />
        </Styled.AddEditor>

        <Styled.AddWidgetsContainer>
          <Styled.AddWidget onClick={focusTagInput}>
            <NotesIcon />
            <TagsInput tags={tags} setTags={setTags} />
          </Styled.AddWidget>

          <AddSubmitButton />
        </Styled.AddWidgetsContainer>
      </form>
    </>
  )
}

export default AddNote
