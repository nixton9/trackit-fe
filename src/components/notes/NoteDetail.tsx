import React, { useState, useEffect } from 'react'
import Tag from './Tag'
import Tooltip from 'react-tooltip-lite'
import { NoteEditor } from './NoteEditor'
import { TagEditor } from './TagEditor'
import { PageLoading } from '../misc/PageLoading'
import { PageError } from '../misc/PageError'
import { NotificationTypes, notificationState } from '../misc/Notification'
import { alertState } from '../misc/Alert'
import { ThreeDotsMenu } from '../misc/ThreeDotsMenu'
import { Styled } from '../../styles/Page.styles'
import { displayDateString, parseDateInverse } from '../../utils/dateHelpers'
import { NoteTag } from '../../utils/ModuleTypes'
import { SINGLE_NOTE } from '../../utils/queries'
import { ReactComponent as ChevronIcon } from '../../assets/icons/chevron.svg'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'
import { ReactComponent as MinusIcon } from '../../assets/icons/minus.svg'
import { Link, RouteComponentProps, useHistory } from 'react-router-dom'
import { useQuery, useMutation, gql } from '@apollo/client'
import { useSetRecoilState } from 'recoil'

const UPDATE_NOTE = gql`
  mutation UpdateNote($id: ID!, $title: String, $content: String) {
    updateNote(id: $id, title: $title, content: $content) {
      id_note
    }
  }
`

const DELETE_NOTE = gql`
  mutation DeleteNote($id: ID!) {
    deleteNote(id: $id) {
      id_note
    }
  }
`

const REMOVE_TAG_FROM_NOTE = gql`
  mutation RemoveTagFromNote($note: ID!, $tag: ID!) {
    removeTagFromNote(note: $note, tag: $tag) {
      id_note
    }
  }
`

type MatchParams = {
  id: string
}

interface MatchProps extends RouteComponentProps<MatchParams> {
  setWidgets: any
}

const NoteDetail: React.FC<MatchProps> = ({ match, setWidgets }) => {
  const [noteTitle, setNoteTitle] = useState('')
  const [noteContent, setNoteContent] = useState('')
  const [showEditor, setShowEditor] = useState(false)
  const [message, setMessage] = useState('')
  const [noteWasEdited, setNoteWasEdited] = useState(false)

  const [showTagEditor, setShowTagEditor] = useState(false)
  const [activeTag, setActiveTag] = useState<NoteTag | null>(null)

  const setNotification = useSetRecoilState(notificationState)
  const setAlert = useSetRecoilState(alertState)

  const { loading, error, data } = useQuery(SINGLE_NOTE, {
    variables: { id: match.params.id }
  })

  const { refetch: refetchNote } = useQuery(SINGLE_NOTE, {
    variables: { id: match.params.id }
  })

  const [updateNote, { error: errorEdit, loading: loadingEdit }] = useMutation(
    UPDATE_NOTE,
    {
      variables: { id: match.params.id, title: noteTitle, content: noteContent }
    }
  )

  const [
    deleteNote,
    { error: errorDelete, loading: loadingDelete }
  ] = useMutation(DELETE_NOTE, {
    variables: {
      id: match.params.id
    }
  })

  const [
    removeTagFromNote,
    { error: errorRemove, loading: loadingRemove }
  ] = useMutation(REMOVE_TAG_FROM_NOTE)

  const history = useHistory()

  const toggleEditor = () => {
    setShowEditor(!showEditor)
  }

  const handleSubmit = () => {
    updateNote()
      .then(res => {
        setNotification({
          text: `Note was successfully updated`,
          type: NotificationTypes.Success
        })
        refetchNote()
      })
      .catch(err =>
        setNotification({
          text: 'There was a problem, please try again',
          type: NotificationTypes.Error
        })
      )
  }

  const handleDeleteConfirm = () => {
    setAlert({
      text: 'This note will be removed.',
      onConfirm: handleDeleteNote
    })
  }

  const handleDeleteNote = () => {
    deleteNote()
      .then(res => {
        history.push(`/notes`)
        setNotification({
          text: `Note was deleted`,
          type: NotificationTypes.Success
        })
      })
      .catch(err => console.log(err.message))
  }

  const handleRemoveTag = (tagId: string | number) => {
    removeTagFromNote({
      variables: {
        note: match.params.id,
        tag: tagId
      }
    })
      .then(res => {
        refetchNote()
        setNotification({
          text: `Tag was removed from note`,
          type: NotificationTypes.Success
        })
      })
      .catch(err =>
        setNotification({
          text: 'There was a problem, please try again',
          type: NotificationTypes.Error
        })
      )
  }

  const handleRemoveTagConfirm = (tagId: string | number) => {
    setAlert({
      text: 'This tag will be removed from this note',
      onConfirm: () => handleRemoveTag(tagId)
    })
  }

  const handleTagClick = (tag: NoteTag) => {
    setActiveTag(tag)
    setShowTagEditor(true)
  }

  const handlePlusClick = () => {
    if (showTagEditor) {
      setShowTagEditor(false)
    } else {
      setActiveTag(null)
      setShowTagEditor(true)
    }
  }

  useEffect(() => {
    setWidgets(false)
    return () => setWidgets(true)
  })

  useEffect(() => {
    if (data) {
      setNoteContent(data.singleNote.content)
      setNoteTitle(data.singleNote.title)
    }
  }, [data])

  useEffect(() => {
    if (data && noteTitle && noteContent) {
      setNoteWasEdited(
        data.singleNote.content !== noteContent ||
          data.singleNote.title !== noteTitle
      )
    }
  }, [data, noteTitle, noteContent])

  useEffect(() => {
    if (message) setTimeout(() => setMessage(''), 1500)
  }, [message])

  const menuOptions = [
    { label: 'Delete note', onClick: handleDeleteConfirm },
    { label: showEditor ? 'Hide editor' : 'Show editor', onClick: toggleEditor }
  ]

  const isLoading = loading || loadingEdit || loadingDelete || loadingRemove

  const errors = error
    ? error
    : errorEdit
    ? errorEdit
    : errorDelete
    ? errorDelete
    : errorRemove
    ? errorRemove
    : null

  return (
    <Styled.PageContainer className="note-detail">
      <Styled.DetailBack>
        <Link to="/notes">
          <ChevronIcon />
        </Link>
      </Styled.DetailBack>

      {errors ? (
        <PageError>{errors.message}</PageError>
      ) : isLoading ? (
        <PageLoading />
      ) : (
        <>
          <Styled.DetailHeader editorActive={showEditor}>
            <div>
              <Styled.DetailTitle
                placeholder="Title for the note"
                value={noteTitle}
                onChange={e => setNoteTitle(e.target.value)}
              />
              <Styled.DetailDate>
                {displayDateString(parseDateInverse(data.singleNote.date))}
              </Styled.DetailDate>
            </div>

            <ThreeDotsMenu options={menuOptions} />
          </Styled.DetailHeader>

          <Styled.DetailTags>
            <Styled.DetailTags__Inner>
              {data.singleNote.tags && data.singleNote.tags.length ? (
                data.singleNote.tags.map((tag: NoteTag) => (
                  <Tooltip
                    eventOff={'onClick'}
                    content={'Click to edit tag'}
                    arrow={false}
                    direction={'up'}
                    className="tag-tooltip"
                  >
                    <Tag
                      key={tag.id}
                      id={tag.id}
                      name={tag.name}
                      color={tag.color}
                      onClick={() => handleTagClick(tag)}
                      onDelete={() => handleRemoveTagConfirm(tag.id)}
                    />
                  </Tooltip>
                ))
              ) : (
                <p>No tags</p>
              )}
            </Styled.DetailTags__Inner>
            <Tooltip
              eventOff={'onClick'}
              content={'Add tag'}
              arrow={false}
              direction={'up'}
            >
              <div onClick={handlePlusClick} className="mbl-click">
                {showTagEditor ? <MinusIcon /> : <PlusIcon />}
              </div>
            </Tooltip>
          </Styled.DetailTags>

          {showTagEditor && (
            <Styled.DetailTagEditor>
              <TagEditor
                tag={activeTag}
                noteId={match.params.id}
                closeEditor={() => setShowTagEditor(false)}
                refetchQuery={refetchNote}
              />
            </Styled.DetailTagEditor>
          )}

          <Styled.DetailContent>
            <NoteEditor
              value={noteContent}
              setValue={setNoteContent}
              showEditor={showEditor}
              readMode
            />
          </Styled.DetailContent>

          {(noteWasEdited || message) && (
            <Styled.DetailSave>
              {message ? (
                <p>{message}</p>
              ) : (
                <Styled.DetailSave__Button
                  onClick={handleSubmit}
                  data-test-id="submit-btn"
                >
                  Save changes
                </Styled.DetailSave__Button>
              )}
            </Styled.DetailSave>
          )}
        </>
      )}
    </Styled.PageContainer>
  )
}

export default NoteDetail
