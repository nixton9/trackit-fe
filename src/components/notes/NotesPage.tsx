import React, { useState } from 'react'
import NotesSettings from './NotesSettings'
import SingleNote from './SingleNote'
import { SelectMenu } from '../misc/SelectMenu'
import { PageLoading } from '../misc/PageLoading'
import { PageError } from '../misc/PageError'
import { activeContentState } from '../misc/Add'
import { Styled } from '../../styles/Page.styles'
import { Note, NoteTag, ModuleTypes } from '../../utils/ModuleTypes'
import { NOTES, TAGS } from '../../utils/queries'
import { SortBySettings } from '../../utils/SettingsTypes'
import { sortData } from '../../utils/globalHelpers'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'
import { ReactComponent as NoDataIcon } from '../../assets/icons/nodata.svg'
import { useQuery } from '@apollo/client'
import { useSetRecoilState } from 'recoil'

const NotesPage: React.FC = () => {
  const setActiveContent = useSetRecoilState(activeContentState)

  const { loading, error, data } = useQuery(NOTES)
  const { data: tags } = useQuery(TAGS, {
    fetchPolicy: 'no-cache'
  })

  const [view, setView] = useState('all')
  const [sortBy, setSortBy] = useState<SortBySettings>(SortBySettings.DATE)

  const viewOptions = tags
    ? [
        { val: 'all', label: 'All' },
        ...tags.tags.map((tag: NoteTag) => ({
          val: tag.id,
          label: tag.name
        }))
      ]
    : [{ val: 'all', label: 'All' }]

  const handleViewChange = (e: React.ChangeEvent<HTMLSelectElement>) =>
    setView(e.target.value)

  const visibleNotes = data
    ? view === 'all'
      ? data.notes
      : data.notes.filter(
          (note: Note) =>
            note.tags?.filter(tag => Number(tag.id) === Number(view)).length
        )
    : []

  const sortedNotes = sortData(visibleNotes, sortBy)

  return (
    <Styled.PageContainer>
      <Styled.PageTitle>Notes</Styled.PageTitle>

      <Styled.PageHeader>
        <Styled.PageHeader__View>
          <Styled.PageHeader__View__Dropdown>
            <SelectMenu
              id="notes-view"
              value={view}
              onChange={handleViewChange}
              options={viewOptions}
              itemClass={'view-select-item'}
            />
          </Styled.PageHeader__View__Dropdown>
          <Styled.PageHeader__View__Counter>
            {data ? data.notes.length : ''}
          </Styled.PageHeader__View__Counter>
        </Styled.PageHeader__View>
        <Styled.PageHeader__Settings>
          <NotesSettings
            tags={tags ? tags.tags : []}
            sortBy={sortBy}
            setSortBy={setSortBy}
          />
        </Styled.PageHeader__Settings>
      </Styled.PageHeader>

      <Styled.PageContent>
        {error ? (
          <PageError>{error.message}</PageError>
        ) : loading ? (
          <PageLoading />
        ) : sortedNotes.length ? (
          (sortedNotes as Note[]).map(note => (
            <SingleNote
              key={note.id}
              id={note.id}
              title={note.title}
              date={note.date}
              tags={note.tags}
            />
          ))
        ) : (
          <Styled.PageContent__NoData>
            <NoDataIcon />
          </Styled.PageContent__NoData>
        )}
      </Styled.PageContent>
      <Styled.PageAddItem onClick={() => setActiveContent(ModuleTypes.Notes)}>
        <PlusIcon />
      </Styled.PageAddItem>
    </Styled.PageContainer>
  )
}

export default NotesPage
