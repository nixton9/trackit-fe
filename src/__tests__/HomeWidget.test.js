import React from 'react'
import { render } from '@testing-library/react'
import HomeWidget, { HomeWidgetTypes } from '../components/misc/HomeWidget'
import { ReactComponent as NotesIcon } from '../assets/icons/notes.svg'
import { BrowserRouter } from 'react-router-dom'

describe('HomeWidget', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <HomeWidget
          type={HomeWidgetTypes.Notes}
          value="32"
          label="created"
          url="/link"
          icon={<NotesIcon />}
        />
      </BrowserRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('has all content', () => {
    const { getByText } = render(
      <BrowserRouter>
        <HomeWidget
          type={HomeWidgetTypes.Notes}
          value="32"
          label="created"
          url="/link"
          icon={<NotesIcon />}
        />
      </BrowserRouter>
    )

    expect(getByText('Notes')).toBeInTheDocument()
    expect(getByText('32')).toBeInTheDocument()
    expect(getByText('created')).toBeInTheDocument()
  })
})
