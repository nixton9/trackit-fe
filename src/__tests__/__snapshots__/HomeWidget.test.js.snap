// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeWidget matches snapshot 1`] = `
<div>
  <a
    href="/notes"
  >
    <div
      class="sc-AxjAm cjhfIZ"
      type="Notes"
    >
      <div
        class="sc-AxirZ faeQuU"
      >
        <h3
          class="sc-AxiKw fHpdZG"
        >
          Notes
        </h3>
        <span
          class="sc-AxhCb fikpvB"
        >
          <svg>
            notes.svg
          </svg>
        </span>
      </div>
      <div
        class="sc-AxhUy bqGvht"
      >
        <h2
          class="sc-AxgMl enJSgo"
        >
          32
        </h2>
        <p
          class="sc-AxheI jctFSf"
        >
          created
        </p>
      </div>
      <span
        class="sc-Axmtr gXbpHN"
      >
        <svg>
          plus.svg
        </svg>
      </span>
    </div>
  </a>
</div>
`;
